### PR TITLE
Initial implementation of Session

### PR DESCRIPTION
**Pull request by TerryHannant**
**Issue**
Resolves #166

**Approach**
Initial implementation of client Session class to encapsulate configuration

### CODE REVIEW

**Comment by ertomatic** [comment]
Can one of the admins verify this patch?

File: ert_shared/client/session.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            self._headers = {"Token": self._connection_info["authtoken"]}
+        else:
+            raise RuntimeError("Invalid Storage Connection configuration")
+
+    def __enter__(self):
+        # Do health check?
+        # Seems redudant
+        return self
+
+    def __exit__(self, *args):
```
Comment (dotfloat) [inline ert_shared/client/session.py:36]:
```suggestion
    def __exit__(self, exc_value: BaseException, exc_type: Type[BaseException], traceback: TracebackType) -> bool:
```

Add imports:
```
from typing import Type
from types import TracebackType
```

No cleanup needed either.

File: ert_shared/client/session.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+        if {"urls", "authtoken"} <= self._connection_info.keys():
+            self._base_url = self.__resolve_url()
+            self._headers = {"Token": self._connection_info["authtoken"]}
+        else:
+            raise RuntimeError("Invalid Storage Connection configuration")
+
+    def __enter__(self):
+        # Do health check?
+        # Seems redudant
```
Comment (dotfloat) [inline ert_shared/client/session.py:33]: Yeah, we can try not using health-checks at this point (we assume that the server is online and operational when the client is invoked).

File: ert_shared/client/session.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        return self
+
+    def __exit__(self, *args):
+        # No real cleanup todo?
+        pass
+
+    def __resolve_url(self) -> str:
```
Comment (dotfloat) [inline ert_shared/client/session.py:40]: We tend to use single underscores for "private" methods
```suggestion
    def _resolve_url(self) -> str:
```

File: ert_shared/client/session.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        # If so is there order of prefence
+        if "ERT_STORAGE_CONNECTION_STRING" in os.environ:
+            try:
+                self._connection_info = json.loads(
+                    os.environ["ERT_STORAGE_CONNECTION_STRING"]
+                )
+            except json.JSONDecodeError:
+                raise RuntimeError("Invalid Storage Connection configuration")
+
+        if {"urls", "authtoken"} <= self._connection_info.keys():
```
Comment (dotfloat) [inline ert_shared/client/session.py:25]: Cool! I haven't seen this pattern before for checking whether a set of keys is contained in a different set. :)

**Review by dotfloat — COMMENTED** [review]
This code should be in https://github.com/equinor/ert-storage I think. Consumers of the ERT Storage API shouldn't rely on the main ERT package. Eg, https://github.com/equinor/webviz-ert doesn't require ERT to operate.

**Comment by dotfloat** [comment]
Oh, there's a decorator in Python called `contextmanager` which lets you define the context manager as a function with a single `yield` rather than having to write a lot of boilerplate: https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager

**Comment by TerryHannant** [comment]
> Oh, there's a decorator in Python called `contextmanager` which lets you define the context manager as a function with a single `yield` rather than having to write a lot of boilerplate: https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager

Yes have used that before, to be honest I was just copying the style used elsewhere in ERT (plugin_manager) in case there was a reason ERT doesn't use them

**Comment by TerryHannant** [comment]
> This code should be in https://github.com/equinor/ert-storage I think. Consumers of the ERT Storage API shouldn't rely on the main ERT package. Eg, https://github.com/equinor/webviz-ert doesn't require ERT to operate.

OK WIll move. I did have it in ert-storage to start with but moved it for some odd reason :-)

File: ert_shared/client/session.py (hunk 3 above)
Comment (TerryHannant) [reply ert_shared/client/session.py:40]: Noted for future

**Comment by TerryHannant** [comment]
Moved to ert-storage

**Pull request closed without merging**